tests/test_parts.py

```diff
@@ -177,6 +177,117 @@ def test_invalid_properties(self):
         assert err[0]["type"] == "value_error.extra"
 
 
+@pytest.fixture
+def charm_exe(tmp_path):
+    """Provide a fake charm executable."""
+    charm_bin = pathlib.Path(tmp_path, "mock_bin", "charm")
+    charm_bin.parent.mkdir(exist_ok=True)
+    charm_bin.write_text(
+        '#!/bin/sh\necho "charmstore-client 2.5.1"\necho "charm-tools version 2.8.2"'
+    )
+    charm_bin.chmod(0o755)
+    yield charm_bin
+
+
+@pytest.fixture
+def broken_charm_exe(tmp_path):
+    """Provide a fake charm executable that fails to run."""
+    charm_bin = pathlib.Path(tmp_path, "mock_bin", "charm")
+    charm_bin.parent.mkdir(exist_ok=True)
+    charm_bin.write_text('#!/bin/sh\nexit 1"')
+    charm_bin.chmod(0o755)
+    yield charm_bin
+
+
+class TestReactivePlugin:
+    """Ensure plugin methods return expected data."""
+
+    @pytest.fixture(autouse=True)
+    def setup_method_fixture(self, tmp_path):
+        project_dirs = craft_parts.ProjectDirs(work_dir=tmp_path)
+        spec = {
+            "plugin": "reactive",
+            "source": str(tmp_path),
+        }
+        plugin_properties = parts.ReactivePluginProperties.unmarshal(spec)
+        part_spec = plugins.extract_part_properties(spec, plugin_name="reactive")
+        part = craft_parts.Part(
+            "foo", part_spec, project_dirs=project_dirs, plugin_properties=plugin_properties
+        )
+        project_info = craft_parts.ProjectInfo(
+            application_name="test",
+            project_dirs=project_dirs,
+            cache_dir=tmp_path,
+            project_name="fake-project",
+        )
+        part_info = craft_parts.PartInfo(project_info=project_info, part=part)
+
+        self._plugin = plugins.get_plugin(
+            part=part,
+            part_info=part_info,
+            properties=plugin_properties,
+        )
+
+    def test_get_build_package(self):
+        assert self._plugin.get_build_packages() == set()
+
+    def test_get_build_snaps(self):
+        assert self._plugin.get_build_snaps() == set()
+
+    def test_get_build_environment(self):
+        assert self._plugin.get_build_environment() == {}
+
+    def test_get_build_commands(self, tmp_path):
+        assert self._plugin.get_build_commands() == [
+            "rm -f charmcraft.yaml",
+            # exit on errors (code 200), warnings (code 100) allowed
+            'charm proof || test "$?" -ge 200 && exit 1',
+            f'mkdir -p "{tmp_path}/parts/foo/install"',
+            f'ln -sf . "{tmp_path}/parts/foo/install/fake-project"',
+            f'charm build -o "{tmp_path}/parts/foo/install" || test "$?" -ge 200 && exit 1',
+            f'rm -f "{tmp_path}/parts/foo/install/fake-project"',
+        ]
```

Comment:
We need to test this better.

This is the core of the plugin, and the test is just an adapted "copy and paste" of above's strings. I'm not worried about current lines (they look fine... but again, they are untested), but this is not useful for future changes, in the sense that if I need to change something in these commands, I would just modify the real code and then "blindly" modify these lines accordingly.

Let's talk about this. I chatted several times with Claudio too about this "limitation" of having "untested bash script lines" as the core of the plugins.

One way of overcoming this limitation is:
- have a `reactive_plugin.py` file.
- that file is imported from `parts.py`, which uses `ReactivePluginEnvironmentValidator`, `ReactivePlugin` etc.
- `get_build_commands` all is does is something like `python3 somepath/reactive_plugin.py`
- `if __name__ == "__main__"` in `reactive_plugin.py` will just run another function
- that function is properly tested

Reply:
https://github.com/canonical/charmcraft/pull/613/commits/b95defda988521dda1d147e60ee940a65892af70 has the same code in a new module, new code upcoming.